File: .conflict-side-0/packages/pstack/src/routing.ts
/**
 * Traefik dynamic configuration files — list, read, write, delete.
 *
 * WHAT THIS IS FOR. Traefik takes configuration from two providers: the **docker** provider (labels
 * on containers — how every per-PR router is declared) and the **file** provider, which watches a
 * directory and reloads on change. The file provider is where everything that is not a container
 * lives: middleware (basicAuth, rate limits, IP allow-lists), TLS options, the catch-all fallback
 * router, and routes to something running outside compose. Splitting it across several files is
 * supported and is the point of a watched *directory* rather than one file — one file per concern,
 * each independently editable.
 *
 * Until now that directory was created by `init` and then never touched again, so the only way to add
 * a middleware was to SSH in. This module is the API behind doing it from the UI.
 *
 * ── WHY THIS NEEDS MORE CARE THAN IT LOOKS ───────────────────────────────────────────────────────
 *
 * Traefik's own documented behaviour: an unparseable file in the watched directory produces a parse
 * error **for the directory**, and the rest of it can be discarded with it. The visible symptom is not
 * "my new middleware is broken", it is *routes elsewhere disappearing*. So a careless write here does
 * not break the thing you were editing — it breaks other people's previews.
 *
 * Three consequences, all load-bearing:
 *
 *   1. **Validate before touching disk.** Parse it, and require the top-level keys to be Traefik's.
 *      A typo'd `htttp:` is valid YAML that configures nothing, and Traefik will not tell you.
 *   2. **Write atomically.** `writeFile` truncates then fills, and the watcher can fire on the
 *      truncated file. Write a temp file in the same directory and `rename` it — rename is atomic
 *      within a filesystem, so the watcher only ever sees a whole file.
 *   3. **Never put anything in that directory that is not dynamic config.** No `.bak`, no temp file
 *      left behind, no dotfile — Traefik would try to parse it. This is why there is no on-disk
 *      version history: the obvious place to keep it is the one place it must not go. `write()`
 *      returns the previous content instead, so a caller can offer an immediate revert.
 *
 * ── WHAT IT CANNOT BREAK ─────────────────────────────────────────────────────────────────────────
 *
 * `control.<domain>` and `api.<domain>` are declared as **docker labels on the pstack container**, not
 * in this directory (see init.ts). So no file written here can lock you out of the UI or the API that
 * would let you fix it. That is not luck; keep it that way.
 */

import { readdir, readFile, rename, rm, stat, writeFile } from 'node:fs/promises';
import { join } from 'node:path';

export class RoutingError extends Error {}

/**
 * A filename that is safe as a path component AND that Traefik will actually read.
 *
 * `.yml`/`.yaml` only: the file provider reads YAML and TOML, but allowing both formats here would
 * mean validating both, and one format is enough. No leading dot (a dotfile is still parsed, and is
 * harder to notice), no directory separator, no `..`.
 */
const NAME = /^[a-z0-9][a-z0-9._-]*\.ya?ml$/;

/** Traefik's top-level dynamic-configuration sections. Anything else is a typo or a mistake. */
const SECTIONS = new Set(['http', 'tcp', 'udp', 'tls']);

export function assertValidRoutingName(name: string): void {
  if (!NAME.test(name) || name.includes('..')) {
    throw new RoutingError(
      `invalid filename "${name}" — must be lowercase, end in .yml or .yaml, and contain no path ` +
        `separators (it becomes a file in Traefik's watched directory)`,
    );
  }
}

/**
 * Reject a file that would break the directory, or silently do nothing.
 *
 * Returns the parsed value on success, so a caller does not parse twice.
 *
 * The "silently does nothing" half matters as much as the parse half: `htttp:` and `middlewares:` at
 * the top level are both perfectly good YAML that Traefik ignores, and an ignored file is a support
 * ticket that starts "I added the middleware and nothing happened".
 */
export function validateRoutingContent(content: string): Record<string, unknown> {
  if (!content.trim()) {
    throw new RoutingError('the file is empty — delete it instead of storing an empty one');
  }

  let parsed: unknown;
  try {
    parsed = Bun.YAML.parse(content);
  } catch (err) {
    // Traefik's failure here is a parse error for the whole DIRECTORY, so this is the check that
    // stops one bad edit taking other routes down with it.
    throw new RoutingError(`not valid YAML: ${(err as Error).message}`);
  }

  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new RoutingError(
      'the top level must be a mapping of Traefik sections, e.g. `http:` with `routers:` under it',
    );
  }

  const keys = Object.keys(parsed as Record<string, unknown>);
  if (keys.length === 0) {
    throw new RoutingError('no sections — expected at least one of http, tcp, udp, tls');
  }
  const unknown = keys.filter((k) => !SECTIONS.has(k));
  if (unknown.length > 0) {
    throw new RoutingError(
      `unknown top-level section(s): ${unknown.join(', ')}. Traefik reads only ` +
        `${[...SECTIONS].join(', ')} — it would load this file and silently apply nothing. ` +
        `Middlewares and routers go *under* \`http:\`.`,
    );
  }

  return parsed as Record<string, unknown>;
}

export type RoutingFile = {
  name: string;
  /** Bytes. Enough for a list view to show something is there without reading every file. */
  size: number;
  /** Epoch milliseconds — mtime. */
  updatedAt: number;
};

/**
 * The watched directory, and whether this process can write to it.
 *
 * `writable` is probed rather than assumed because the pstack container only gained this mount in
 * 0.4.0: on a host whose control stack predates it the directory is simply absent from the
 * container, and the API must say "re-run `pstack init`" rather than fail with ENOENT.
 */
export class RoutingStore {
  readonly dir: string;

  constructor(dir: string) {
    this.dir = dir;
  }

  /** True when the directory exists and this process may write in it. */
  async writable(): Promise<boolean> {
    try {
      const s = await stat(this.dir);
      if (!s.isDirectory()) return false;
      // Probe by writing, not by reading a mode bit: the mount may be present but `:ro`, which is
      // exactly the misconfiguration worth detecting, and a mode check would not catch it.
      const probe = join(this.dir, `.pstack-write-probe-${process.pid}`);
      await writeFile(probe, '', 'utf8');
      await rm(probe, { force: true });
      return true;
    } catch {
      return false;
    }
  }

  async list(): Promise<RoutingFile[]> {
    let entries: string[];
    try {
      entries = await readdir(this.dir);
    } catch {
      return [];
    }
    const files: RoutingFile[] = [];
    for (const name of entries) {
      if (!NAME.test(name)) continue; // never present a file Traefik would not read
      try {
        const s = await stat(join(this.dir, name));
        if (s.isFile()) files.push({ name, size: s.size, updatedAt: Math.trunc(s.mtimeMs) });
      } catch {
        /* vanished between readdir and stat — nothing to report */
      }
    }
    return files.sort((a, b) => a.name.localeCompare(b.name));
  }

  async read(name: string): Promise<string> {
    assertValidRoutingName(name);
    try {
      return await readFile(join(this.dir, name), 'utf8');
    } catch {
      throw new RoutingError(`no such routing file: ${name}`);
    }
  }

  /**
   * Validate, then replace atomically. Returns the previous content, or null if it is new.
   *
   * The temp file goes in the SAME directory because `rename` is only atomic within a filesystem —
   * across one it degrades to copy-then-delete, which is the partial-file window this exists to
   * avoid. It is named so that Traefik will not read it (no `.yml` suffix) and is removed on failure,
   * because a leftover file in this directory is exactly what must never happen.
   */
  async write(name: string, content: string): Promise<string | null> {
    assertValidRoutingName(name);
    validateRoutingContent(content);
    if (!(await this.writable())) {
      throw new RoutingError(
        `Traefik's dynamic directory is not writable from here (${this.dir}). The control stack ` +
          `mounts it into the API from 0.4.0 onward — re-run \`pstack init\` on the host to pick ` +
          `up the mount.`,
      );
    }

    const target = join(this.dir, name);
    const previous = await readFile(target, 'utf8').catch(() => null);
    const tmp = join(this.dir, `.pstack-tmp-${process.pid}-${name}.part`);
    try {
      await writeFile(tmp, content, 'utf8');
      await rename(tmp, target);
    } catch (err) {
      await rm(tmp, { force: true }).catch(() => {});
      throw new RoutingError(`could not write ${name}: ${(err as Error).message}`);
    }
    return previous;
  }

  /** Returns the removed content, so a caller can offer an undo. */
  async remove(name: string): Promise<string> {
    assertValidRoutingName(name);
    const content = await this.read(name);
    if (!(await this.writable())) {
      throw new RoutingError(`Traefik's dynamic directory is not writable from here (${this.dir}).`);
    }
    await rm(join(this.dir, name), { force: true });
    return content;
  }
}

File: packages/pstack/internal/routing/routing.go
// Package routing is Traefik dynamic configuration files — list, read, write, delete.
//
// WHAT THIS IS FOR. Traefik takes configuration from two providers: the **docker** provider (labels
// on containers — how every per-PR router is declared) and the **file** provider, which watches a
// directory and reloads on change. The file provider is where everything that is not a container
// lives: middleware (basicAuth, rate limits, IP allow-lists), TLS options, the catch-all fallback
// router, and routes to something running outside compose. Splitting it across several files is
// supported and is the point of a watched *directory* rather than one file — one file per concern,
// each independently editable.
//
// Until now that directory was created by `init` and then never touched again, so the only way to
// add a middleware was to SSH in. This module is the API behind doing it from the UI.
//
// ── WHY THIS NEEDS MORE CARE THAN IT LOOKS ───────────────────────────────────────────────────────
//
// Traefik's own documented behaviour: an unparseable file in the watched directory produces a parse
// error **for the directory**, and the rest of it can be discarded with it. The visible symptom is
// not "my new middleware is broken", it is *routes elsewhere disappearing*. So a careless write here
// does not break the thing you were editing — it breaks other people's previews.
//
// Three consequences, all load-bearing:
//
//  1. **Validate before touching disk.** Parse it, and require the top-level keys to be Traefik's.
//     A typo'd `htttp:` is valid YAML that configures nothing, and Traefik will not tell you.
//  2. **Write atomically.** A plain write truncates then fills, and the watcher can fire on the
//     truncated file. Write a temp file in the same directory and `rename` it — rename is atomic
//     within a filesystem, so the watcher only ever sees a whole file.
//  3. **Never put anything in that directory that is not dynamic config.** No `.bak`, no temp file
//     left behind, no dotfile — Traefik would try to parse it. This is why there is no on-disk
//     version history: the obvious place to keep it is the one place it must not go. `Write()`
//     returns the previous content instead, so a caller can offer an immediate revert.
//
// ── WHAT IT CANNOT BREAK ─────────────────────────────────────────────────────────────────────────
//
// `control.<domain>` and `api.<domain>` are declared as **docker labels on the pstack container**,
// not in this directory (see initctl). So no file written here can lock you out of the UI or the
// API that would let you fix it. That is not luck; keep it that way.
package routing

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"

	"github.com/samishal1998/preview-stacks/packages/pstack/internal/omap"
	"github.com/samishal1998/preview-stacks/packages/pstack/internal/yamlx"
)

// Error is a refused name, refused content, an unwritable directory or a missing file. The API
// maps it to 400 with its own sentence.
type Error struct{ Msg string }

func (e *Error) Error() string { return e.Msg }

// IsError reports whether err is a *Error.
func IsError(err error) bool {
	var e *Error
	return errors.As(err, &e)
}

// A filename that is safe as a path component AND that Traefik will actually read.
//
// `.yml`/`.yaml` only: the file provider reads YAML and TOML, but allowing both formats here would
// mean validating both, and one format is enough. No leading dot (a dotfile is still parsed, and is
// harder to notice), no directory separator, no `..`.
var nameRe = regexp.MustCompile(`^[a-z0-9][a-z0-9._-]*\.ya?ml$`)

// sections are Traefik's top-level dynamic-configuration sections, in the order the message lists
// them. Anything else is a typo or a mistake.
var sections = []string{"http", "tcp", "udp", "tls"}

// AssertValidRoutingName rejects a filename Traefik would not read or that could leave the directory.
func AssertValidRoutingName(name string) error {
	if !nameRe.MatchString(name) || strings.Contains(name, "..") {
		return &Error{fmt.Sprintf(`invalid filename "%s" — must be lowercase, end in .yml or .yaml, and contain no path separators (it becomes a file in Traefik's watched directory)`, name)}
	}
	return nil
}

// ValidateRoutingContent rejects a file that would break the directory, or silently do nothing.
//
// Returns the parsed mapping on success, so a caller does not parse twice.
//
// The "silently does nothing" half matters as much as the parse half: `htttp:` and `middlewares:`
// at the top level are both perfectly good YAML that Traefik ignores, and an ignored file is a
// support ticket that starts "I added the middleware and nothing happened".
func ValidateRoutingContent(content string) (*omap.Map, error) {
	if strings.TrimSpace(content) == "" {
		return nil, &Error{"the file is empty — delete it instead of storing an empty one"}
	}
	parsed, err := yamlx.ParseString(content)
	if err != nil {
		// Traefik's failure here is a parse error for the whole DIRECTORY, so this is the check that
		// stops one bad edit taking other routes down with it.
		return nil, &Error{"not valid YAML: " + strings.TrimPrefix(err.Error(), "not valid YAML: ")}
	}
	doc, ok := parsed.(*omap.Map)
	if !ok {
		return nil, &Error{"the top level must be a mapping of Traefik sections, e.g. `http:` with `routers:` under it"}
	}
	keys := doc.Keys()
	if len(keys) == 0 {
		return nil, &Error{"no sections — expected at least one of http, tcp, udp, tls"}
	}
	var unknown []string
	for _, k := range keys {
		if !isSection(k) {
			unknown = append(unknown, k)
		}
	}
	if len(unknown) > 0 {
		return nil, &Error{fmt.Sprintf("unknown top-level section(s): %s. Traefik reads only %s — it would load this file and silently apply nothing. Middlewares and routers go *under* `http:`.", strings.Join(unknown, ", "), strings.Join(sections, ", "))}
	}
	return doc, nil
}

func isSection(k string) bool {
	for _, s := range sections {
		if s == k {
			return true
		}
	}
	return false
}

// RoutingFile is one entry of the list view.
type RoutingFile struct {
	Name string `json:"name"`
	// Size in bytes. Enough for a list view to show something is there without reading every file.
	Size int64 `json:"size"`
	// UpdatedAt is epoch milliseconds — mtime.
	UpdatedAt int64 `json:"updatedAt"`
}

// RoutingStore is the watched directory, and whether this process can write to it.
//
// Writable is probed rather than assumed because the pstack container only gained this mount in
// 0.4.0: on a host whose control stack predates it the directory is simply absent from the
// container, and the API must say "re-run `pstack init`" rather than fail with ENOENT.
//
// Stateless apart from Dir; safe to share. No lock: two writers race on the rename, and the last
// whole file wins — the reference had the same property.
type RoutingStore struct {
	Dir string
	// wildcardMu serialises the wildcard's multi-file writes — see SetWildcard. It guards nothing
	// else: the single-file routes keep the last-whole-file-wins property described above.
	wildcardMu sync.Mutex
}

// New returns the store over dir.
func New(dir string) *RoutingStore { return &RoutingStore{Dir: dir} }

// Writable is true when the directory exists and this process may write in it.
func (s *RoutingStore) Writable() bool {
	st, err := os.Stat(s.Dir)
	if err != nil || !st.IsDir() {
		return false
	}
	// Probe by writing, not by reading a mode bit: the mount may be present but `:ro`, which is
	// exactly the misconfiguration worth detecting, and a mode check would not catch it.
	probe := filepath.Join(s.Dir, fmt.Sprintf(".pstack-write-probe-%d", os.Getpid()))
	if err := os.WriteFile(probe, nil, 0o666); err != nil {
		return false
	}
	_ = os.Remove(probe)
	return true
}

// List is every file Traefik would read, by name.
func (s *RoutingStore) List() []RoutingFile {
	files := []RoutingFile{}
	entries, err := os.ReadDir(s.Dir)
	if err != nil {
		return files
	}
	for _, e := range entries {
		name := e.Name()
		if !nameRe.MatchString(name) {
			continue // never present a file Traefik would not read
		}
		st, err := os.Stat(filepath.Join(s.Dir, name))
		if err != nil {
			continue // vanished between readdir and stat — nothing to report
		}
		if st.Mode().IsRegular() {
			files = append(files, RoutingFile{Name: name, Size: st.Size(), UpdatedAt: st.ModTime().UnixMilli()})
		}
	}
	// Byte order where the reference used localeCompare — the same thing for the lowercase ASCII
	// the name rule admits.
	sort.SliceStable(files, func(i, j int) bool { return files[i].Name < files[j].Name })
	return files
}

// Read returns a file's content.
func (s *RoutingStore) Read(name string) (string, error) {
	if err := AssertValidRoutingName(name); err != nil {
		return "", err
	}
	b, err := os.ReadFile(filepath.Join(s.Dir, name))
	if err != nil {
		return "", &Error{"no such routing file: " + name}
	}
	return string(b), nil
}

// Write validates, then replaces atomically. Returns the previous content, or nil if it is new.
//
// The temp file goes in the SAME directory because `rename` is only atomic within a filesystem —
// across one it degrades to copy-then-delete, which is the partial-file window this exists to
// avoid. It is named so that Traefik will not read it (no `.yml` suffix) and is removed on failure,
// because a leftover file in this directory is exactly what must never happen.
func (s *RoutingStore) Write(name, content string) (*string, error) {
	if err := assertNotReserved(name); err != nil {
		return nil, err
	}
	return s.write(name, content)
}

// write is Write without the reserved-name gate — for the routes that OWN a reserved file.
func (s *RoutingStore) write(name, content string) (*string, error) {
	if err := AssertValidRoutingName(name); err != nil {
		return nil, err
	}
	if _, err := ValidateRoutingContent(content); err != nil {
		return nil, err
	}
	if !s.Writable() {
		return nil, &Error{fmt.Sprintf("Traefik's dynamic directory is not writable from here (%s). The control stack mounts it into the API from 0.4.0 onward — re-run `pstack init` on the host to pick up the mount.", s.Dir)}
	}
	target := filepath.Join(s.Dir, name)
	var previous *string
	if b, err := os.ReadFile(target); err == nil {
		p := string(b)
		previous = &p
	}
	tmp := filepath.Join(s.Dir, fmt.Sprintf(".pstack-tmp-%d-%s.part", os.Getpid(), name))
	if err := os.WriteFile(tmp, []byte(content), 0o666); err != nil {
		_ = os.Remove(tmp)
		return nil, &Error{fmt.Sprintf("could not write %s: %s", name, err)}
	}
	if err := os.Rename(tmp, target); err != nil {
		_ = os.Remove(tmp)
		return nil, &Error{fmt.Sprintf("could not write %s: %s", name, err)}
	}
	return previous, nil
}

// Remove deletes a file and returns the removed content, so a caller can offer an undo.
func (s *RoutingStore) Remove(name string) (string, error) {
	if err := assertNotReserved(name); err != nil {
		return "", err
	}
	return s.remove(name)
}

// remove is Remove without the reserved-name gate — for the routes that OWN a reserved file.
func (s *RoutingStore) remove(name string) (string, error) {
	if err := AssertValidRoutingName(name); err != nil {
		return "", err
	}
	content, err := s.Read(name)
	if err != nil {
		return "", err
	}
	if !s.Writable() {
		return "", &Error{fmt.Sprintf("Traefik's dynamic directory is not writable from here (%s).", s.Dir)}
	}
	if err := os.Remove(filepath.Join(s.Dir, name)); err != nil && !errors.Is(err, os.ErrNotExist) {
		return "", err
	}
	return content, nil
}
